docs/clients/dotnet/21.2/migration-to-gRPC.md

```diff
@@ -0,0 +1,599 @@
+
+`StreamPosition` class has an overloaded operator for `ulong` value assignment. 
+
+```csharp
+StreamPosition streamPosition = 100L;
+```
+
+You can also convert it from the `StreamRevision`:
+
+```csharp
+var streamPosition = StreamPosition.FromStreamRevision(streamRevision);
+```
+
+For reading from `$all`, the TCP Client has already the `Position` class. gRPC client also has it, but using `ulong` instead of `long` for commit and prepare positions.
+
+### Read result
+
+TCP Client requires paging through the results. You must provide the maximum number of events you want to read in the single read call. In the gRPC client, this is optional. By default, it will try to read all events. To make it efficient, read methods return [IAsyncEnumerable](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8). Which means that it won't load the whole stream at once but iterate sequentially, reducing the memory pressure.
+
+Instead of doing paging like that in TCP client:
+
+```csharp
+public async Task<IEnumerable<object>> LoadEvents(string stream)
+{
+    const int pageSize = 4096;
+
+    var start  = 0;
+    var events = new List<object>();
+
+    do
+    {
+        var page = await _connection.ReadStreamEventsForwardAsync(
+            stream, start, pageSize, true
+        );
+
+        if (page.Status == SliceReadStatus.StreamNotFound)
+            throw new ArgumentOutOfRangeException(
+                nameof(stream), $"Stream '{stream}' was not found"
+            );
+
+        events.AddRange(
+            page.Events.Select(Deserialize)
+        );
+        if (page.IsEndOfStream) break;
+
+        start += pageSize;
+    } while (true);
+
+    return events;
+}
+```
+
+You can simplify that in the gRPC client into:
+
+```csharp
+public async Task<IEnumerable<object>> LoadEvents(string stream)
+{
+    await using var readResult = _client.ReadStreamAsync(
+        Direction.Forwards,
+        stream,
+        StreamPosition.Start
+    );
+
+    if(await readResult.ReadState != ReadState.Ok)
+        throw new ArgumentOutOfRangeException(
+            nameof(stream), $"Stream '{stream}' was not found"
+        );
+
+    return await readResult
+        .Select(Deserialize)
+        .ToListAsync();
+}
+```
+
+### Serialisation
+
+The gRPC client uses `ReadOnlyMemory<byte>` instead of `byte` array to make the events processing more efficient. To support that, you need to modify your deserialisation logic slightly:
+
+```csharp{4-5}
+  object Deserialize(this ResolvedEvent resolvedEvent)
+  {
+      var dataType = TypeMapper.GetType(resolvedEvent.Event.EventType);
+-     var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data);
++     var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data.Span);
+      var data = JsonConvert.DeserializeObject(jsonData, dataType);
```

Comment:
Do we need to use Newtonsoft.Json in these examples? .NET serializer can directly serialize to bytes.

Reply:
Added a disclaimer in https://github.com/EventStore/documentation/pull/486/commits/3e9d2d2757645583da3718b5b3ec8eef2c983165.